### ISSUE TITLE
Add support for generic attributes

### ISSUE THREAD

**Issue by luisvt** [dart-lang/reflectable#85]
Let say I have next class:

``` dart
class Person {
  String id;

  List<Address> addresses;
}
```

then I need to get, by mean of reflection, the type of the list addresses which is `Address` at client/browser side

**Comment by fromlabs** [comment]
I need this too. I've implemented a small DI library ([dartregistry](https://pub.dartlang.org/packages/dartregistry)) and I did a porting from mirrors to reflectable a few months ago. Type arguments would be very helpful for writing code like this:
```dart
class MyService {
  @inject
  // actually a use this annotation @Inject(SecurityService)
  Provider<SecurityService> securityService;
}
```
Unfortunately the client/browser transformation throws this error:

> Exception: UnimplementedError: *** Unfortunately, this feature has not yet been implemented: typeArguments.
> If you wish to ensure that it is prioritized, please report it on github.com/dart-lang/reflectable.

**Comment by eernstg** [comment]
Thanks for the input! I won't get a chance to look into this right now, but I'll try to squeeze it in during the coming weeks. Note that the case where the type argument is given statically as in the type annotation `List<Address>` may be within reach, but the case where it is a type parameter (if it had been `List<E>` inside `Person<E>`) will still be blocked on the lack of suitable primitives, so it won't be a complete solution.

**Comment by jolleekin** [comment]
Will this feature be implemented anytime soon, at least for the static case?

**Comment by eernstg** [comment]
I guess this ought to teach me not to predict what's going to happen during the coming weeks. ;-)

Anyway, the basic problem still exists: We have no way to generate code at compile time that does not modify the target class (and we can't modify corelib classes like `List` and its implementations) which will allow us to query the type arguments of an existing instance or `Type`. We simply do not have the primitives needed to decompose types at run time.

This means that we cannot extract the type argument for a given instance of `List<T>` for some unknown `T` (and class covariance ensures that this type argument is almost always unknown), nor from a given `Type` value which is known to represent `List<T>` for some unknown `T`, nor for `X` in the following:
```dart
class MyService<X extends SecurityService> {
  @inject
  Provider<X> securityService;
}
```
So when you say 'static case' it's not enough that we are querying a type annotation, it is also required that said type annotation does not depend on any type variables from its enclosing class.

The example you give satisfies these constraints,
```dart
class MyService {
  @inject
  // actually a use this annotation @Inject(SecurityService)
  Provider<SecurityService> securityService;
}
```
so we should be able to generate code for creating a representation of the type argument `SecurityService` and for similar "simple" cases. I've illustrated how something like that could be achieved in [this gist](https://gist.github.com/eernstg/15b6b8590691f66877e7cfe25efb97d1), which runs the following code with output 'SecurityService':
```dart
import 'package:reflectable/reflectable.dart';
import 'main.reflectable.dart';

class Reflector extends Reflectable {
  const Reflector() : super(instanceInvokeCapability, declarationsCapability,
      typeCapability, typeRelationsCapability, reflectedTypeCapability);
}

const Reflector reflector = const Reflector();

@reflector
class SecurityService {}

@reflector
class Provider<X> {}

@reflector
class MyService {
  Provider<SecurityService> securityService;
}

main() {
  initializeReflectable();
  ClassMirror myServiceMirror = reflector.reflectType(MyService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.declarations;
  VariableMirror securityServiceMirror = declarations['securityService'];
  ClassMirror typeAnnotationMirror = securityServiceMirror.type;
  List<Type> reflectedTypeArguments =
      typeAnnotationMirror.reflectedTypeArguments;
  print(reflectedTypeArguments[0]);
}
```
It's highly incomplete, though. For instance, it breaks if an actual type argument contains a type variable declared by an enclosing scope. But it gives a hint about how to do this.

**Comment by eernstg** [comment]
OK, this is one step better, it now detects nested occurrences of type variables (deeming the given type unsupported, which means that we get an `UnimplementedError` when trying to get the `reflectedTypeArguments`):
```dart
import 'package:reflectable/reflectable.dart';
import 'main.reflectable.dart';

class Reflector extends Reflectable {
  const Reflector() : super(instanceInvokeCapability, declarationsCapability,
      typeCapability, typeRelationsCapability, reflectedTypeCapability);
}

const Reflector reflector = const Reflector();

@reflector
class SecurityService {}

@reflector
class Provider<X> {}

@reflector
class MyService {
  Provider<SecurityService> securityService;
}

@reflector
class MyGenericService<X extends SecurityService> {
  Provider<X> genericSecurityService;
}

plainTest() {
  ClassMirror myServiceMirror = reflector.reflectType(MyService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.declarations;
  VariableMirror securityServiceMirror = declarations['securityService'];
  ClassMirror typeAnnotationMirror = securityServiceMirror.type;
  List<Type> reflectedTypeArguments =
      typeAnnotationMirror.reflectedTypeArguments;
  print(reflectedTypeArguments[0]);
}

genericTest() {
  ClassMirror myServiceMirror = reflector.reflectType(MyGenericService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.declarations;
  VariableMirror securityServiceMirror =
      declarations['genericSecurityService'];
  ClassMirror typeAnnotationMirror = securityServiceMirror.type;
  try {
    typeAnnotationMirror.reflectedTypeArguments;
  } on UnimplementedError catch (_) {
    print("OK: not implemented.");
  }
}

main() {
  initializeReflectable();
  plainTest();
  genericTest();
}
```
The [gist](https://gist.github.com/eernstg/15b6b8590691f66877e7cfe25efb97d1) has been updated with the improved code.

**Comment by eernstg** [comment]
I think I can clean up the `reflectedTypeArguments` patch I mentioned and put it into reflectable as the next thing, after branch `use_build` is ready and has become the next version of reflectable. It should be more robust and it should reuse types in the database, rather than relying on generated literal type expressions, but the interface seen by clients of reflectable should be stable.

**Comment by jolleekin** [comment]
Thanks for the great work. I can't wait to use it in my packages.

**Comment by eernstg** [comment]
Here's a preview: Use branch `support_static_type_arguments_nov17`. Check out [`static_type_arguments_test.dart`](https://github.com/dart-lang/reflectable/blob/support_static_type_arguments_nov17/test_reflectable/test/static_type_arguments_test.dart).

**Comment by eernstg** [comment]
Reflectable has been updated to remove all traces of transformers, so branch `support_static_type_arguments_nov17` is completely out of date. Created a new one at `support_static_type_arguments_aug18` which works with reflectable 2.0.4 and which is now under review.

**Comment by luisvt** [comment]
@fromlabs I recommend you to test [built_mirrors](https://pub.dartlang.org/packages/built_mirrors), it already has support for generics, for example:

```dart
library built_mirrors_generic_sample;

import 'package:built_mirrors/built_mirrors.dart';

part 'main.g.dart';

@reflectable
class SecurityService {}

@reflectable
class Provider<X> {}

@reflectable
class MyService {
  Provider<SecurityService> securityService;
}

@reflectable
class MyGenericService<X extends SecurityService> {
  Provider<X> genericSecurityService;
}

plainTest() {
  ClassMirror myServiceMirror = reflectType(MyService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.fields;
  DeclarationMirror securityServiceMirror = declarations['securityService'];
  List<Type> types = securityServiceMirror.type;
  print(types[0]); // instead of printing Provicer<T> will print Provider<dynamic>
  print(types[1]); // after reflection since type is T in fact is a dynamic
}

genericTest() {
  ClassMirror myServiceMirror = reflectType(MyGenericService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.fields;
  DeclarationMirror securityServiceMirror = declarations['genericSecurityService'];
  List<Type> types = securityServiceMirror.type;
  print(types[0]); // instead of printing Provicer<T> will print Provider<dynamic>
  print(types[1]); // after reflection since type is T in fact is a dynamic
}

main() {
  _initMirrors();
  plainTest();
  genericTest();
}
```

**Comment by eernstg** [comment]
That's cool, @luisvt!

Actually, reflectable will do something very similar already today:
```dart
library reflectable_generic_sample;

import 'package:reflectable/reflectable.dart';
import 'main.reflectable.dart';

class MyReflectable extends Reflectable {
  const MyReflectable(): super(instanceInvokeCapability,
      typeRelationsCapability, reflectedTypeCapability);
}

const reflectable = MyReflectable();

@reflectable
class SecurityService {}

@reflectable
class Provider<X> {}

@reflectable
class MyService {
  Provider<SecurityService> securityService;
}

@reflectable
class MyGenericService<X extends SecurityService> {
  Provider<X> genericSecurityService;
}

plainTest() {
  ClassMirror myServiceMirror = reflectable.reflectType(MyService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.declarations;
  VariableMirror securityServiceMirror = declarations['securityService'];
  Type type = securityServiceMirror.reflectedType;
  print(type); // Prints 'Provider<SecurityService>'.

  Type typeArgument =
      securityServiceMirror.type.typeArguments[0].reflectedType;
  print(typeArgument); // Prints 'SecurityService'.
  print(typeArgument == SecurityService); // Prints 'true'.
}

genericTest() {
  ClassMirror myServiceMirror = reflectable.reflectType(MyGenericService);
  Map<String, DeclarationMirror> declarations = myServiceMirror.declarations;
  VariableMirror securityServiceMirror =
      declarations['genericSecurityService'];
  Type type = securityServiceMirror.reflectedType;
  
  // The `type` is synthetic because we cannot deliver a representation of
  // `Provider<X>` for a class: For an instance of `MyGenericService<int>`
  // it would have to be the `Type` for `Provider<int>`, and for a
  // `MyGenericService<String>` it must be `Provider<String>`. But here
  // we are just exploring the class, so `X` has no specific value.
  print(type); // Prints `Type(reflectable_generic_sample.Provider<X>)`.

  // But we can still explore it and find an approximation: The "dynamic
  // reflected type". Further explanation given below.
  ClassMirror classMirror = securityServiceMirror.type;
  print(classMirror.hasDynamicReflectedType); // Prints 'true'.
  print(classMirror.dynamicReflectedType); // Prints 'Provider<dynamic>'.
}

main() {
  initializeReflectable();
  plainTest();
  genericTest();
}
```

We are able to obtain a `Type` for the type annotation of a field (and similar things like the return type of a method and so on), and we can print it and see that it is a parameterized type like `Provider<SecurityService>` when the field is `MyService.securityService`.

In the case where we explore a generic class like `MyGenericService<X>` we'd have to deliver different values for the type annotation of `genericSecurityService`, for each different value of `X`. In other words, it's not a meaningful request to say "give me the `Type` for that type annotation of the class", it would have to be for a specific generic instantiation of the class. For any given instance it would have the type `MyGenericService<T>` for some specific `T`, and the correct `Type` for the type annotation would then print as `Provider<T>`. But that's exactly where we do not have the primitives needed in order to extract the value of `T` from the type of any given instance.

So we can't deliver such a `Type` for the class because it makes no sense, and we cannot deliver it for the class of any given instance, because we don't have the required primitives.

I can see that you've chosen to make it `Provider<dynamic>`.

This is actually also supported by reflectable, but it has a different name because it isn't the actual value of that type annotation "for the class" or for any instance, it is an approximation.

It's called the `dynamicReflectedType`, which is basically `MyType<dynamic, dynamic, ...>` where the actual type is `MyType<T1, ..., Tk>` for some types `Tj`, and you can use that in the case where the request is meaningless (because we are reflecting upon a class) or unavailable (because we are reflecting on a generic instantiation of the generic class, e.g., based on an instance), and it will be a common supertype for all the possible actual values that we would have gotten, if it were possible.

Using the dynamic reflected type reflectable will deliver the `Type` for `Provider<dynamic>`, just like `built_mirrors`. But if we change the declaration of `Provider` such that `X` has a bound:

```dart
class Provider<X extends SecurityService> {}
```

then we'll get `Provider<SecurityService>` instead, that is, we get the result from [instantiate-to-bound](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/instantiate-to-bound.md), which will be a supertype of all the possible values that we can't obtain directly.

Finally, I included four lines of code at the end of `plainTest` illustrating which new features [the type argument PR](https://github.com/dart-lang/reflectable/pull/148) adds. They fail at run time with reflectable 2.0.4 ("not yet supported"), but they are working as described when that PR has landed.

**Comment by eernstg** [comment]
[2.0.5](https://pub.dartlang.org/packages/reflectable) was just landed and published. This means that the example shown [above](https://github.com/dart-lang/reflectable/issues/85#issuecomment-416906428) will work with a regular reflectable dependency and the PR and branch can be ignored.

Also, the [build.yaml](https://github.com/dart-lang/reflectable/blob/master/build.yaml) settings were made more inclusive, such that a plain `pub run build_runner build` will also generate code for libraries in `bin` and `example` and other directories which as "standard entrypoint directories" for pub.

**Comment by eernstg** [comment]
@luisvt, I can see how `<Type>[List, Person]` may be a reasonable representation for `List<Person>`, but it is actually ambiguous: If you receive a result of the form `[T1, T2, T3]`, how do you know whether `T1` is a generic class that takes two type arguments (meaning `T1<T2, T3>`, which could be something like `Map<int, String>`, which you could confirm with `T1 == Map`) or `T1` is a generic class taking one type argument and so is `T2` (meaning `T1<T2<T3>>`, which could be `List<List<int>>`)?

Or maybe you'll make it a `List<dynamic>` and then you will use `[T1, [T2, T3]]` for the latter?

Anyway, it looks like there is a need to deal with those ambiguities no matter which representation you'll use.

**Comment by eernstg** [comment]
@luisvt, I had to go do other things, here's my response to some more of what you wrote:

> main problem .. is .. how to construct those generics

Right, this is again something that we would need additional primitives in order to do, and it may well be the case that programs get a lot bigger and/or slower if such primitives are made available.

In particular, you can't express a generic instantiation of a class (like `MyGeneric<Person>`) based on values of type `Type` (like a `Type` for `MyGeneric<dynamic>` and another one for `Person`), and that is not even supported by 'dart:mirrors' (which is a platform library, that is, it could in principle add its own primitives as needed because it is part of the "system", unlike anything which is based on code generation). So we can't even get hold of a `Type` for such a dynamically constructed parameterized type. Your program would have to have a way to express `MyGeneric<Person>` directly in regular (non-reflective) code, and then you would be able to get hold of the corresponding `Type`.

Next, there is no way in Dart to use a `Type` to create a new instance, short of using reflection to get a mirror for that type (which would not be able to access the specific values of type arguments if we're talking reflection based on generated code), or coding up something like a `Map` that maps instances of `Type` to some factory functions. Those factory functions would have to be written such that each of them just calls a specific constructor with some type arguments that must be available at the call site as types:

```dart
class MyGeneric<X extends num, Y extends X> {}

Map<Type, Function> factories = {
  List: <X>() => <X>[],
  MyGeneric: <X extends num, Y extends X>() => MyGeneric<X, Y>(),
};

main() {
  Type someType = MyGeneric;
  var x = (factories[someType])<num, int>();
  print(x.runtimeType); // Prints 'MyGeneric<num, int>'.
}
```

So you can actually create instances of a class which is chosen dynamically and pass actual type arguments which are expressed as a type (it won't work with a `Type` instance), but you have to write all those factory functions manually.

Of course, we might want to support something like that using code generation, but I'm not quite sure it's worth the effort, because of the inherent limitations in the expressive power.

**Comment by luisvt** [comment]
the ambiguity is eliminate if you check if the second argument is an array, for example:

```
T1 -> T1
T1<T2> -> [T1, [T2]]
T1<T2, T3> -> [T1, [T2, T3]]
T1<T2<T3>> -> [T1, [T2, [T3]]]
T1<T2<T3, T4>> -> [T1, [T2, [T2, T4]]]
```

**Comment by eernstg** [comment]
OK, that's just because I wasn't sure whether to interpret `[Map, [String, Person]` as `[Map, String, Person]` or `[Map, [String, Person]]`.

However, I don't think it makes such a big difference as long as you can get hold of the actual type arguments of a given parameterized type (that is, that you can obtain `int` from `List<int>`), and that's possible with the list-of-lists based representation as well as with `TypeMirror.typeArguments` resp. `TypeMirror.reflectedTypeArguments`. That works fine for statically resolved types.

What we still don't have, however, with any approach using generated code, is the ability to extract a type argument which isn't statically resolved. For instance, with the following `C`:

```dart
class C<X> { List<X> xs; }
```
.. you could have an instance with dynamic type `C<Foo>` and static type `C<dynamic>` (or some other applicable type argument, that doesn't actually matter), but you wouldn't be able to use any `...Mirror` to re-discover that it is actually a `C` of `Foo`, and you wouldn't be able to discover (using a `DeclarationMirror` etc.) that its `xs` has a type annotation whose actual type is `List<Foo>`, because we simply don't have the primitives needed to get that information.

**Comment by eernstg** [comment]
Right, but it's worth noting that 2 is undecidable (we cannot hope to enumerate all possible actual type argument lists at compile-time). For example:
```dart
class A<X> {
  A();
  A<A<X>> get next => A<A<X>>();
}

main() {
  var i = 20;
  A a = A<Null>();
  while (--i > 0) a = a.next;
  print(a.runtimeType);
}
```
will print `A<A<A<A<A<A<A<A<A<A<A<A<A<A<A<A<A<A<A<A<Null>>>>>>>>>>>>>>>>>>>>`.

**Comment by eernstg** [comment]
Closing: I've decided to close this issue because we've taken the steps that we can take without obtaining new special privileges (new primitives). So it's considered to be closed by commit 55a67a47ced32554893cef26a1b10096f51aba17.

A new issue #151 has been created to handle a missing case (obtaining the actual type arguments of an existing instance), and others may be created later if some reasonably implementable and useful approaches come up.
